File: src/csharp/pulse/cdm/properties/SEScalarVolumePerPressure.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class VolumePerPressureUnit : Unit
  {
    private VolumePerPressureUnit(string v) : base(v) { }

    public static VolumePerPressureUnit L_Per_Pa = new VolumePerPressureUnit("L/Pa");
    public static VolumePerPressureUnit L_Per_cmH2O = new VolumePerPressureUnit("L/cmH2O");
    public static VolumePerPressureUnit mL_Per_cmH2O = new VolumePerPressureUnit("mL/cmH2O");
    public static VolumePerPressureUnit m3_Per_Pa = new VolumePerPressureUnit("m^3/Pa");
    public static VolumePerPressureUnit mL_Per_mmHg = new VolumePerPressureUnit("mL/mmHg");

    public static VolumePerPressureUnit FromString(string u)
    {
      if (u == VolumePerPressureUnit.L_Per_Pa.ToString())
        return VolumePerPressureUnit.L_Per_Pa;
      if (u == VolumePerPressureUnit.L_Per_cmH2O.ToString())
        return VolumePerPressureUnit.L_Per_cmH2O;
      if (u == VolumePerPressureUnit.mL_Per_cmH2O.ToString())
        return VolumePerPressureUnit.mL_Per_cmH2O;
      if (u == VolumePerPressureUnit.m3_Per_Pa.ToString())
        return VolumePerPressureUnit.m3_Per_Pa;
      if (u == VolumePerPressureUnit.mL_Per_mmHg.ToString())
        return VolumePerPressureUnit.mL_Per_mmHg;

      throw new System.ArgumentException(u + " is not a know volume per  pressure unit");
    }
  }

  public class SEScalarVolumePerPressure : SEScalar
  {
    protected VolumePerPressureUnit unit;

    public SEScalarVolumePerPressure() : base()
    {

    }

    public SEScalarVolumePerPressure(SEScalarVolumePerPressure from)
    {
      Set(from);
    }

    public SEScalarVolumePerPressure(double value, VolumePerPressureUnit unit)
    {
      SetValue(value, unit);
    }

    public void Set(SEScalarVolumePerPressure from)
    {
      if (from == null)
      {
        Invalidate();
        return;
      }
      SetValue(from.value, from.unit);
    }

    public new void SetValue(double value)
    {
      throw new System.NotSupportedException("You must provide a unit");
    }

    public void SetValue(double value, VolumePerPressureUnit unit)
    {
      this.value = value;
      this.unit = unit;
    }

    public double GetValue(VolumePerPressureUnit unit)
    {
      return UnitConverter.Convert(this.value, this.unit, unit);
    }

    public VolumePerPressureUnit GetUnit()
    {
      return unit;
    }

    public new string ToString()
    {
      return this.value + "(" + this.unit.ToString() + ")";
    }
  }
}

File: src/cpp/cdm/properties/SEScalarVolumePerPressure.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/properties/SEScalarVolumePerPressure.h"

const VolumePerPressureUnit VolumePerPressureUnit::L_Per_Pa("L/Pa");
const VolumePerPressureUnit VolumePerPressureUnit::L_Per_cmH2O("L/cmH2O");
const VolumePerPressureUnit VolumePerPressureUnit::mL_Per_cmH2O("mL/cmH2O");
const VolumePerPressureUnit VolumePerPressureUnit::m3_Per_Pa("m^3/Pa");
const VolumePerPressureUnit VolumePerPressureUnit::mL_Per_mmHg("mL/mmHg");

bool VolumePerPressureUnit::IsValidUnit(const std::string& unit)
{
  if (L_Per_Pa.GetString().compare(unit) == 0)
    return true;
  if (L_Per_cmH2O.GetString().compare(unit) == 0)
    return true;
  if (mL_Per_cmH2O.GetString().compare(unit) == 0)
    return true;
  if (m3_Per_Pa.GetString().compare(unit) == 0)
    return true;
  if (mL_Per_mmHg.GetString().compare(unit) == 0)
    return true;
  return false;
}

const VolumePerPressureUnit& VolumePerPressureUnit::GetCompoundUnit(const std::string& unit)
{
  if (L_Per_Pa.GetString().compare(unit) == 0)
    return L_Per_Pa;
  if (L_Per_cmH2O.GetString().compare(unit) == 0)
    return L_Per_cmH2O;
  if (mL_Per_cmH2O.GetString().compare(unit) == 0)
    return mL_Per_cmH2O;
  if (m3_Per_Pa.GetString().compare(unit) == 0)
    return m3_Per_Pa;
  if (mL_Per_mmHg.GetString().compare(unit) == 0)
    return mL_Per_mmHg;
  throw CommonDataModelException(unit + " is not a valid VolumePerPressure unit");
}

template class SEScalarQuantity<VolumePerPressureUnit>;
